src/main/java/net/minecraftforge/event/terraingen/ChunkGeneratorEvent.java

```diff
@@ -106,4 +114,195 @@ public InitNoiseField(IChunkGenerator chunkProvider, double[] noisefield, int po
         public int getSizeY() { return sizeY; }
         public int getSizeZ() { return sizeZ; }
     }
+
+    /**
+     * This event is fired during chunk generation as well as structure regeneration before any of the vanilla structure gens are executed.
+     * If fired during recreateStructures the chunkprimer will be null
+     * This event is fired regardless of mapFeaturesEnabled.
+     */
+    public static class Generate extends ChunkGeneratorEvent
+    {
+        private final World world;
+        private final @Nullable ChunkPrimer primer;
+        private final int x, z;
+
+        public Generate(IChunkGenerator gen, World world, int x, int z, @Nullable ChunkPrimer primer)
+        {
+            super(gen);
+            this.world = world;
+            this.x = x;
+            this.z = z;
+            this.primer = primer;
+        }
+
+        public World getWorld()
+        {
+            return world;
+        }
+
+        public @Nullable ChunkPrimer getPrimer()
+        {
+            return primer;
+        }
+
+        public int getX()
+        {
+            return x;
+        }
+
+        public int getZ()
+        {
+            return z;
+        }
+    }
+
+    /**
+     * This event is fired when the chunk generator is asked for a list of possible creatures for a specific location.
+     * (Vanilla checks if there is a structure which wants to spawn special creatures at the location)
+     * <p>
+     * To return your own spawn entries, add them to the list and cancel this event.
```

Comment:
I'm a little rusty on making new events but is this usually how it's done?

I thought usually you just expose some list for the modder to modify, no cancellation is required.

Reply:
The thing is, the vanilla Minecraft code only allows one structure to return it's list and defaults to the biome's spawn list if there is no relevant structure. 
Pseudocode:
```
if(structureA)
  return structureA.getCreatures()
if(structureB)
  return structureB.getCreatures()
return biome.getCreatures
```
This does not really fit to the usual event style of just adding everything together (vanilla and mod), which would not require cancellation.
To keep the patch size simple I just stayed with the vanilla style. If the even is cancelled the event's creature list is returned otherwise the vanilla code continues. I could check the size of the list instead of the cancellation status, but this would not allow enforcing an empty spawn list.